### PR TITLE
DRAFT: Add New (Optional) Non-Differential Quorum Algorithm

### PR DESCRIPTION
**Pull request by Bill**
There are problems with Geode's current quorum algorithm. In particular, split-brains are possible if locators are started at the same time, or if they are started during a network partition (and can't reach other locators).

This (draft) PR introduces an alternative quorum maintenance algorithm, activated by a new optional configuration parameter `quorum-absolute-locator-count`. When it's configured to a value greater-than-zero, two things are done differently:

1. when a locator is starting or restarting, and it might potentially become the coordinator, it can become the coordinator only if it has communicated with `quorum-absolute-locator-count` locators (including itself).

2. when a coordinator is considering whether a view change might result in a loss of quorum, rather than basing that decision on the change in membership weight between the current view and the prospective one, the locator count in the prospective view is compared to `quorum-absolute-locator-count`. If the locator count is below the threshold then it's treated as lost quorum, with all the usual processing that happens in that event.

The ability to switch between the legacy quorum algorithm and this new one is accomplished via a new policy object. `GMSJoinLeave` has a reference to a `Quorum` object. Legacy behavior is captured in the `DifferentialQuorum` implementation, and the new algorithm is embodied in the `AbsoluteQuorum` implementation. `GMSJoinLeave` delegates to the `Quorum` object in the `GMSJoinLeave.isLostQuorum()` method. (That method used to be called `isNetworkPartition()` but it was renamed since quorum can be lost due to reasons other than a network partition.)

In the legacy quorum algorithm, no quorum checking is done at cluster startup (when a coordinator is chosen). The new algorithm has to operate there. As a result `GMSJoinLeave.findCoordinator()` was modified to delegate to the `Quorum` object (so that the new algorithm can prevent cluster formation if too few locators are currently reachable).

This PR is very drafty so be kind.

### Checklist:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

**Comment by Bill** [comment]
Those three latest failures in distributed tests (2 in JDK8, one in JDK11) all have open tickets so they are not caused by work on this branch.

File: geode-core/src/test/java/org/apache/geode/distributed/internal/DistributionConfigJUnitTest.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ public void before() {
   @Test
   public void testGetAttributeNames() {
     String[] attNames = AbstractDistributionConfig._getAttNames();
-    assertThat(attNames.length).isEqualTo(169);
+    assertThat(attNames.length).isEqualTo(170);
```
Comment (kirklund) [inline geode-core/src/test/java/org/apache/geode/distributed/internal/DistributionConfigJUnitTest.java:107]: This syntax provides better failure messages:
```
assertThat(attNames).hasSize(170);
```
I'd go ahead and convert the test to use only AssertJ. There's an IntelliJ plugin that can automate this but it's not very smart, I usually end up having to touch up anything involving collections, arrays, or messages.

**Review by kirklund — COMMENTED** [review]
I like the approach you have in this PR. Are you still working on it (it's in Draft)? I just recommend some small assertion changes.

**Comment by onichols-pivotal** [comment]
this PR appears to be abandoned, can it be closed?

**Review by onichols-pivotal — CHANGES_REQUESTED** [review]
_comment deleted_

**Review by onichols-pivotal — APPROVED** [review]

**Comment by metatype** [comment]
Closing this out due to age.

**Pull request closed without merging**